openvdb_ax/openvdb_ax/codegen/StandardFunctions.cc

```diff
@@ -0,0 +1,1982 @@
+        return std::max(a, b);
+    };
+
+    return FunctionBuilder("max")
+        .addSignature<double(double,double)>(generate, (double(*)(double,double))(max))
+        .addSignature<float(float,float)>(generate, (float(*)(float,float))(max))
+        .addSignature<int32_t(int32_t,int32_t)>(generate, (int32_t(*)(int32_t,int32_t))(max))
+        .setArgumentNames({"a", "b"})
+        .addFunctionAttribute(llvm::Attribute::ReadOnly)
+        .addFunctionAttribute(llvm::Attribute::NoRecurse)
+        .addFunctionAttribute(llvm::Attribute::NoUnwind)
+        .addFunctionAttribute(llvm::Attribute::AlwaysInline)
+        .setConstantFold(op.mConstantFoldCBindings)
+        .setPreferredImpl(op.mPrioritiseIR ? FunctionBuilder::IR : FunctionBuilder::C)
+        .setDocumentation("Returns the larger of the given values.")
+        .get();
+}
+
+inline FunctionGroup::Ptr axclamp(const FunctionOptions& op)
+{
+    auto generate =
+        [op](const std::vector<llvm::Value*>& args,
+             llvm::IRBuilder<>& B) -> llvm::Value*
+    {
+        llvm::Value* min = axmax(op)->execute({args[0], args[1]}, B);
+        llvm::Value* result = axmin(op)->execute({min, args[2]}, B);
+        return result;
+    };
+
+    using ClampD = double(double, double, double);
+    using ClampF = float(float, float, float);
+    using ClampI = int32_t(int32_t, int32_t, int32_t);
+
+    return FunctionBuilder("clamp")
+        .addSignature<ClampD>(generate, &openvdb::math::Clamp<double>)
+        .addSignature<ClampF>(generate, &openvdb::math::Clamp<float>)
+        .addSignature<ClampI>(generate, &openvdb::math::Clamp<int32_t>)
+        .addDependency("min")
+        .addDependency("max")
+        .addFunctionAttribute(llvm::Attribute::ReadOnly)
+        .addFunctionAttribute(llvm::Attribute::NoRecurse)
+        .addFunctionAttribute(llvm::Attribute::NoUnwind)
+        .addFunctionAttribute(llvm::Attribute::AlwaysInline)
+        .setArgumentNames({"in", "min", "max"})
+        .setConstantFold(op.mConstantFoldCBindings)
+        .setPreferredImpl(op.mPrioritiseIR ? FunctionBuilder::IR : FunctionBuilder::C)
+        .setDocumentation("Clamps the first argument to the minimum second argument "
+            "value and maximum third argument value")
+        .get();
+}
+
+inline FunctionGroup::Ptr axfit(const FunctionOptions& op)
+{
+    auto generate =
+        [op](const std::vector<llvm::Value*>& args,
+             llvm::IRBuilder<>& B) -> llvm::Value*
+    {
+        //         (outMax - outMin)(x - inMin)
+        // f(x) = ----------------------------  + outMin
+        //                inMax - inMin
+        // if inMax == inMin, f(x) = (outMax + outMin) / 2.0
+
+        // NOTE: this also performs a clamp on the ordered input range
+        // @TODO revisit. If this is the best thing to do, should add conditional
+        // branching so that the clamping math is never executed when the value
+        // is inside
```

Comment:
Input range should not be forced to be ordered.
Clamping the input range makes sense; VEX uses efit() for the case you don't want clamping.

Here's the VEX definition:
``` 
    float d = omax - omin;
    float tmp = select(d == 0, (nmin + nmax) * 0.5,
            nmin + (nmax - nmin) * (val - omin) / d);
    return select(nmin < nmax,
            clamp(tmp, nmin, nmax),
            clamp(tmp, nmax, nmin));
```

n is New, o is Old, so will look a bit backwards.

Reply:
I believe our version is the same as this, but it clamps `val` to `omin` and `omax` first rather than the result to `nmin` and `nmax`. Something like

```
val =  select(omin< omax,  clamp(val , omin, omax),  clamp(val , omax, omin));
float d = omax - omin;
float tmp = select(d == 0, (nmin + nmax) * 0.5,
            nmin + (nmax - nmin) * (val - omin) / d);
```